### PR TITLE
[Merged by Bors] -  Implement product image selection

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

For https://github.com/stackabletech/operator-rs/issues/470
Reference implementation of https://github.com/stackabletech/operator-rs/pull/476

## Review Checklist

- [ ] Code contains useful comments
- [ ] CRD change approved (or not applicable)
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by sbernauer** [comment]
Started test-run https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/kafka-operator-it-custom/21/

**Review by teozkr — APPROVED** [review]
Overall this looks like a nice ergonomics improvement on the operator side

File: rust/crd/Cargo.toml [hunk 1]
```diff
@@ -9,7 +9,7 @@ version = "0.8.0-nightly"
 publish = false
 
 [dependencies]
-stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", tag = "0.26.0" }
+stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", branch = "feature/product-image-selection" }
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/crd/Cargo.toml:12]: thx!

File: rust/operator/src/discovery.rs [hunk 2]
```diff
@@ -95,9 +100,7 @@ fn build_discovery_configmap(
                 .with_recommended_labels(
                     kafka,
                     APP_NAME,
-                    kafka
-                        .image_version()
-                        .context(KafkaVersionParseFailureSnafu)?,
+                    &resolved_product_image.product_version,
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator/src/discovery.rs:101]: done

File: rust/operator/src/kafka_controller.rs [hunk 3]
```diff
@@ -444,9 +454,7 @@ fn build_broker_role_serviceaccount(
             .with_recommended_labels(
                 kafka,
                 APP_NAME,
-                kafka
-                    .image_version()
-                    .context(KafkaVersionParseFailureSnafu)?,
+                &resolved_product_image.product_version,
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:454]: This change implies that we now one have different label values. Previously we had e.g. `1.4.1-stackable2.1.0`, now we have `1.4.1`. I'm not entirely sure if this causes problems, but in the `custom` product image variant the stackable version isn't known, so we can't add the "fully qualified" version.

File: rust/operator/src/kafka_controller.rs [hunk 4]
```diff
@@ -242,16 +243,16 @@ impl ReconcilerError for Error {
 }
 
 pub async fn reconcile_kafka(kafka: Arc<KafkaCluster>, ctx: Arc<Ctx>) -> Result<Action> {
+    let resolved_product_image = kafka.spec.image.resolve(DOCKER_IMAGE_BASE_NAME);
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator/src/kafka_controller.rs]: No reason. Moved it down

File: rust/operator/src/kafka_controller.rs (hunk 3 above)
Comment (sbernauer) [reply rust/operator/src/kafka_controller.rs:454]: Had a short talk at arch meeting and decided to wait on the discussion regarding calling our product-images e.g. `1.4.1-release22.09`

File: rust/operator/src/kafka_controller.rs [hunk 5]
```diff
@@ -409,9 +419,7 @@ pub fn build_broker_role_service(kafka: &KafkaCluster) -> Result<Service> {
             .with_recommended_labels(build_recommended_labels(
                 kafka,
                 KAFKA_CONTROLLER_NAME,
-                kafka
-                    .image_version()
-                    .context(KafkaVersionParseFailureSnafu)?,
+                &resolved_product_image.product_version,
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:422]:
```suggestion
                &resolved_product_image.app_version_label,
```

File: rust/operator/src/kafka_controller.rs [hunk 6]
```diff
@@ -441,9 +450,7 @@ fn build_broker_role_serviceaccount(
...
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:453]:
```suggestion
                &resolved_product_image.app_version_label,
```

File: rust/operator/src/kafka_controller.rs [hunk 7]
```diff
@@ -460,9 +467,7 @@ fn build_broker_role_serviceaccount(
...
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:470]:
```suggestion
                &resolved_product_image.app_version_label,
```

File: rust/operator/src/kafka_controller.rs [hunk 8]
```diff
@@ -506,9 +512,7 @@ fn build_broker_rolegroup_config_map(
                 .with_recommended_labels(build_recommended_labels(
                     kafka,
                     KAFKA_CONTROLLER_NAME,
-                    kafka
-                        .image_version()
-                        .context(KafkaVersionParseFailureSnafu)?,
+                    &resolved_product_image.product_version,
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:515]:
```suggestion
                    &resolved_product_image.app_version_label,
```

File: rust/operator/src/kafka_controller.rs [hunk 9]
```diff
@@ -548,9 +553,7 @@ fn build_broker_rolegroup_service(
             .with_recommended_labels(build_recommended_labels(
                 kafka,
                 KAFKA_CONTROLLER_NAME,
-                kafka
-                    .image_version()
-                    .context(KafkaVersionParseFailureSnafu)?,
+                &resolved_product_image.product_version,
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:556]:
```suggestion
                &resolved_product_image.app_version_label,
```

File: rust/operator/src/kafka_controller.rs [hunk 10]
```diff
@@ -834,12 +835,13 @@ fn build_broker_rolegroup_statefulset(
             m.with_recommended_labels(build_recommended_labels(
                 kafka,
                 KAFKA_CONTROLLER_NAME,
-                image_version,
+                &resolved_product_image.product_version,
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:838]:
```suggestion
                &resolved_product_image.app_version_label,
```

File: rust/operator/src/kafka_controller.rs [hunk 11]
```diff
@@ -882,7 +884,7 @@ fn build_broker_rolegroup_statefulset(
             .with_recommended_labels(build_recommended_labels(
                 kafka,
                 KAFKA_CONTROLLER_NAME,
-                image_version,
+                &resolved_product_image.product_version,
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:887]:
```suggestion
                &resolved_product_image.app_version_label,
```

**Review by sbernauer — COMMENTED** [review]
LGTM otherwise

**Comment by sbernauer** [comment]
LGTM when tests pass

**Comment by maltesander** [comment]
test https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/kafka-operator-it-custom/26/

**Pull request closed without merging**